SourceCode/include/tracer/RayTracer.h

```diff
@@ -2,28 +2,40 @@
 #include <string>
 #include <vector>
 
-#include "Image.h"
 #include "Ray.h"
-#include "SceneParser.h"
-#include "Triangle.h"
+#include "SceneSettings.h"
 #include "tracer/Camera.h"
+#include "tracer/Triangle.h"
+#include "tracer/Vector.h"
+
+typedef Vector ColorVector;
+#define MAX_DEPTH 5
 
 class RayTracer {
  private:
+  struct IntersectionInformation {
+    Vector intersectionPoint;
+    const Mesh *const object;
+    const Triangle *const triangle;
+  };
+
   bool rayUpdateRequired;
-  Image image;
-  Camera camera;
-  SceneParser sceneParser;
-  std::vector<Triangle> sceneTriangles;
-  std::vector<std::vector<Ray>> pixelRays;
+  bool renderRequired;
+  Scene scene;
+  std::vector<std::vector<Ray<Primary>>> pixelRays;
+  std::vector<std::vector<Color>> colorBuffer;
   void updateRays();
+  Color shootRay(const Ray<Primary> &ray, const unsigned int depth = 0) const;
+  template <RayType T>
+  Color shade(const Ray<T> &ray) const;
+  template <RayType T>
+  std::optional<RayTracer::IntersectionInformation> trace(const Ray<T> &ray) const;
+  bool hasIntersection(const Ray<Shadow> &ray) const;
 
  public:
-  RayTracer(const Image &image, const Camera &camera, const std::string &pathToScene);
-  const Image &getImage() const;
-  Image &setImage();
+  explicit RayTracer(const std::string &pathToScene);
   const Camera &getCamera() const;
   Camera &setCamera();
-  void updateScene(const std::string &pathToNewScene);
-  void generatePPM(const std::string &pathToImage);
+  void render();
+  void writePPM(const std::string &pathToImage);
```

Comment:
This class is getting quite heavy, consider decoupling the logic soon. The RayTracer/Render should be responsible only for the tracing